embassy-nrf/src/uarte.rs

```diff
@@ -0,0 +1,335 @@
+//! Async low power UARTE.
+//!
+//! The peripheral is autmatically enabled and disabled as required to save power.
+//! Lowest power consumption can only be guaranteed if the send receive futures
+//! are dropped correctly (e.g. not using `mem::forget()`).
+
+use core::future::Future;
+use core::sync::atomic::{compiler_fence, Ordering};
+use core::task::{Context, Poll};
+
+use embassy::util::Signal;
+use embedded_dma::{StaticReadBuffer, StaticWriteBuffer};
+
+use crate::fmt::{assert, *};
+use crate::hal::gpio::Port as GpioPort;
+use crate::hal::pac;
+use crate::hal::prelude::*;
+use crate::hal::target_constants::EASY_DMA_SIZE;
+use crate::interrupt;
+use crate::interrupt::OwnedInterrupt;
+
+pub use crate::hal::uarte::Pins;
+// Re-export SVD variants to allow user to directly set values.
+pub use pac::uarte0::{baudrate::BAUDRATE_A as Baudrate, config::PARITY_A as Parity};
+
+/// Interface to the UARTE peripheral
+pub struct Uarte {
+    instance: pac::UARTE0,
+    _irq: interrupt::UARTE0_UART0Interrupt,
+}
+
+struct State {
+    tx_done: Signal<()>,
+    rx_done: Signal<u32>,
+}
+
+static STATE: State = State {
+    tx_done: Signal::new(),
+    rx_done: Signal::new(),
+};
+
+// TODO: Remove when https://github.com/nrf-rs/nrf-hal/pull/276 has landed
+#[cfg(any(feature = "52833", feature = "52840"))]
+fn port_bit(port: GpioPort) -> bool {
+    match port {
+        GpioPort::Port0 => false,
+        GpioPort::Port1 => true,
+    }
+}
+
+impl Uarte {
+    pub fn new(uarte: pac::UARTE0, irq: interrupt::UARTE0_UART0Interrupt) -> Self {
+        // Enable interrupts
+        uarte.events_endtx.reset();
+        uarte.events_endrx.reset();
+        uarte
+            .intenset
+            .write(|w| w.endtx().set().txstopped().set().endrx().set().rxto().set());
+        // TODO: Set interrupt priority?
```

Comment:
Some other drivers used to set a hardcoded priority, I've removed that. 
User can set the priority on the irq before passing it to us if they care about it.